### ISSUE TITLE
simplify.resolve blows up call stack in case of recursive references

### ISSUE THREAD

**Issue by gwhitney** [josdejong/mathjs#2383]
Simplest example:
```
const expr = math.parse('a')
math.simplify.resolve(expr, {a:expr})
```

Silghtly more subtle:
```
const asubexpr = math.parse('a+1')
const bsubexpr = math.parse('b+2')
math.simplify.resolve(asubexpr, {a:bsubexpr, b:asubexpr})
```

Both calls throw  `RangeError: Maximum call stack size exceeded`.

It seems to me that any of the following would be more civilized behavior:

(a) Return the original expression unchanged.

(b) Only substitute a given variable once down any branch of the parse tree. In other words, the second snippet would return `a+1+2+1` in that it would substitute `b+2` into `a+1` to get `b+2+1`, and then substitute `a+1` into that to get `a+1+2+1` and then refuse to substitute `a` again because it had already substituted `a` farther up that branch of the parse tree.

(c) Detect such loops and throw a mathjs-specific error object with some diagnostic information (and not take the time to fill up the entire call stack).

With some guidance on which sort of resolution would be most in tune with mathjs architecture and that this seems worth fixing (it can be subtle in a complicated scope whether a reference loop has been created), I'd be happy to take a stab at a PR.

**Comment by josdejong** [comment]
Yes, good point, `simplify` cannot handle circular dependencies right now.

I doubt if solving a circular dependency once (option b) really helps the user in general. A circular dependency cannot be solve like this, so it requires the user to make a change. So then we have roughly two options: (a) do nothing, let the user wonder why nothing is is happening and figure out the circular dependency, or (c) throw an exception (or maybe allow returning both a result and a warning message). Do you have any preference?

**Comment by gwhitney** [comment]
I don't really like option (a) myself actually, I was trying to be complete -- I agree it leaves the user wondering. So it seems one of the variations of (c) is the way to go. I think throwing an error will likely be less problematic here -- clients of the code might just ignore any auxiliary message, but throwing will force them to catch the error and make a conscious decision how to handle. I will put in the docs for resolve that it might throw, so as to minimize surprises. Do you have a preference as to the type of the error? I was thinking a `ReferenceError` with a message like "recursive loop of variable definitions" together with a list of the variables in the loop. So in all, something like
```
ReferenceError: recursive loop of variable definitions a -> b -> z -> a
```

**Comment by josdejong** [comment]
I like option (c) the most too, let's go for that approach then, throwing an error when a circular dependency is detected👍 

Your proposal to implement it as a `ReferenceError` with a clear message sounds good. A message explaining the full circular path like `a -> b -> z -> a` would be best. However, if it is very complex to achieve that full trace, I think it may be ok to be pragmatic and just return limited information like it is related to variable `a`.

**Comment by gwhitney** [comment]
Hmmm, so far as I can see there is no documentation of `simplify.resolve` -- so I suppose the PR should supply some? Also, should it appear in the documentation as `math.simplify.resolve` or just `simplify.resolve`? I am a bit confused as to why on the `docs/reference/functions.html` page, the only functions that appear without the "math." prefix are derivative, rationalize, and simplify. Is there something different about those? Or should I file a (very minor) issue about the discrepancy in the documentation of those?

**Comment by josdejong** [comment]
It can very well be that the docs are not consistent in that regard. To me it makes sense to not mention the `math.` prefix, since this prefix is not always there depending on how you use mathjs. You can do `import * as mathjs from 'mathjs'`, resulting in a `mathjs` prefix, or `import { simplify } from 'mathjs'`, resulting in no prefix at all. Does that make sense.

I also see that there there is indeed no documentation for functions like `simplify.resolve`. I think they originally started as internal functions, then needed to be exposed, but never got proper documentation. Adding docs would be great, but feel free to pick this up as part of this PR or not, parking it for later.

**Comment by josdejong** [comment]
The fix for this (#2405) is published now in `v10.2.0`. Thanks again!
